src/seeds/types.ts

```diff
@@ -0,0 +1,143 @@
+        () => minimumKeysPresent(payload, PlatformValidKeys),
+        () => onlyValidKeysPresent(payload, PlatformValidKeys),
+        () => typeValidator(payload, 'type', MusicPlatformType),
+      ],
+      'update': [
+        () => minimumKeysPresent(payload, ['id']),
+        () => onlyValidKeysPresent(payload, PlatformValidKeys),
+        () => typeValidatorOptional(payload, 'type', MusicPlatformType),
+      ],
+    },
+    'nftFactories': {
+      'upsert': [
+        () => minimumKeysPresent(payload, NFTFactoryMinUpsertKeys),
+        () => onlyValidKeysPresent(payload, NFTFactoryValidUpsertKeys),
+        () => typeValidator(payload, 'contractType', NFTContractTypeName),
+        () => typeValidator(payload, 'standard', NFTStandard),
+      ],
+      'update': [
+        () => minimumKeysPresent(payload, ['id']),
+        () => onlyValidKeysPresent(payload, NFTFactoryValidUpdateKeys),
+      ],
+    },
+    'artists': {
+      'upsert': [
+        () => minimumKeysPresent(payload, ArtistValidKeys),
+        () => onlyValidKeysPresent(payload, ArtistValidKeys),
+      ],
+      'update': [
+        () => minimumKeysPresent(payload, ['id']),
+        () => onlyValidKeysPresent(payload, ArtistValidKeys),
+      ],
+    },
+    'processedTracks': {
+      'upsert': [
+        () => minimumKeysPresent(payload, ProcessedTrackValidKeys),
+        () => onlyValidKeysPresent(payload, ProcessedTrackValidKeys),
+      ],
+      'update': [
+        () => minimumKeysPresent(payload, ['id']),
+        () => onlyValidKeysPresent(payload, ProcessedTrackValidKeys),
+      ],
+    }
+  }
+
+  validatorFunctions[payload.entity][payload.operation].forEach((fn: any) => fn());
+}
+
+const minimumKeysPresent = (input: SeedPayload, keys: any): void => {
+  if (!keys.every((key: any) => input.data.hasOwnProperty(key))) {
+    throw new Error(`${input.entity} entity is missing required fields`)
+  }
+}
+
+const onlyValidKeysPresent = (input: SeedPayload, keys: any): void => {
+  if (!Object.keys(input.data).every((key: any) => keys.includes(key))) {
+    throw new Error(`${input.entity} entity has unsupported fields`)
+  }
+}
+
+const typeValidator = (input: SeedPayload, key: string, validOptions: any): void => {
+  if (!Object.values(validOptions).includes(input.data[key])) {
+    throw new Error(`not a valid ${input.entity} ${key}`)
+  }
+}
+
+const typeValidatorOptional = (input: SeedPayload, key: string, validOptions: any): void => {
+  if (input.data[key]) {
+    typeValidator(input, key, validOptions);
+  }
+}
+
+const entityValidator = (input: SeedPayload): void => {
+  if (!Object.values(SeedEntities).includes(input.entity)) {
+    throw new Error('unknown seed entity');
+  }
+  if (!Object.values(CrdtOperation).includes(input.operation)) {
+    throw new Error('must specify either `upsert` or `update` operation');
+  }
+}
+
+export const persistSeed = async (payload: SeedPayload) => {
+  let dbClient: any;
+  entityValidator(payload);
```

Comment:
Strictly speaking no as we've run it with `validateSeed`.  

But I do like the idea of independently validating the inputs for both the exported functions in case they're used outside of the API.

Reply:
yeh same - should we do the full validation here too then not just entityValidator?

Reply:
Yep, great solution. Adjusted and pushed up 👍